feat: Starknet messaging function and event reference (based on #1094) (#1209)
starknet-io/starknet-docs: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc

[id="messaging_function_reference"]
= Messaging function and event reference
:description: Comprehensive function and event reference for the `StarknetMessaging` smart contracts.

> The `StarknetMessaging` smart contracts include functions and events that you use to implement various messaging flows in a dApp.

Author: @LucasLvy Can you suggest what you think should be here?

For information on the messaging mechanism, see xref:architecture_and_concepts:Network_Architecture/messaging-mechanism.adoc[Messaging mechanism].

The L1 functions, where available, are defined in the following smart contract:

[cols=",,",]
|===
|Contract |Description |Functions

|https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`] |
The contract that implements messaging functions.

Interfaces are available through https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessaging.sol[`IStarknetMessaging.sol`]. a|
* xref:#cancelL1ToL2Message[`cancelL1ToL2Message`]
* xref:#consumeMessageFromL2[`consumeMessageFromL2`]
* xref:#getMaxL1MsgFee[`getMaxL1MsgFee`]
* xref:#l1ToL2Messages[`l1ToL2Messages`]
* xref:#sendMessageToL2[`sendMessageToL2`]

> * xref:#startL1ToL2MessageCancellation[`startL1ToL2MessageCancellation`]

Reviewer: these are only the functions defined in `IStarknetMessaging.sol` it misses the functions implemented in `StarknetMessaging.sol` which are not part of the interface.

|===

The L1 events, where available, are defined in the following smart contract:

[cols=",,",]
|===
|Contract |Description |Events

|https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol[`IStarknetMessagingEvents.sol`] |
The contract that defines messaging events. a|
* xref:#ConsumedMessageToL1[`ConsumedMessageToL1`]
* xref:#ConsumedMessageToL2[`ConsumedMessageToL2`]
* xref:#LogMessageToL1[`LogMessageToL1`]
* xref:#LogMessageToL2[`LogMessageToL2`]
* xref:#MessageToL2Canceled[`MessageToL2Canceled`]
* xref:#MessageToL2CancellationStarted[`MessageToL2CancellationStarted`]

|===

== L1 function reference

Functions are listed in alphabetical order.

'''

[#cancelL1ToL2Message]
=== `cancelL1ToL2Message`

[discrete]
==== Description

> Cancels an L1 to L2 message. Call this function after calling the xref:#startL1ToL2MessageCancellation[`startL1ToL2MessageCancellation`] function. The time between the calls to these two functions must be at least the number of seconds defined by the link:https://github.com/starkware-libs/cairo-lang/blob/caba294d82eeeccc3d86a158adb8ba209bf2d8fc/src/starkware/starknet/solidity/StarknetMessaging.sol#L25[`L1L2_MESSAGE_CANCELLATION_DELAY_TAG`] constant. #I can't find an actual value for this constant.#

Reviewer: Sort of correct, the constant defines the storage slot the value is actually stored at. To get the real value one should call [this](https://github.com/starkware-libs/cairo-lang/blob/caba294d82eeeccc3d86a158adb8ba209bf2d8fc/src/starkware/starknet/solidity/StarknetMessaging.sol#L59-L61) function.

Reviewer: Adding a link to the function on etherscan would be great also so it's easy to see the current value.

Reviewer: https://etherscan.io/address/0xc662c410c0ecf747543f5ba90660f6abebd9c8c4#readProxyContract#F11

`uint256 _toAddress_`:: The address of the L2 contract.
`uint256 _selector_`::
include::messaging_reference.adoc[tag=uint256_selector]
`uint256[] calldata _payload_`:: The payload of the message.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]
`bytes32 _msgHash_`:: The hash of the message.
`uint256 _nonce_`:: The nonce of the message.

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L110[`sendMessageToL2`]

'''

[#startL1ToL2MessageCancellation]
=== `startL1ToL2MessageCancellation`

[discrete]
==== Description

Starts the cancellation of an L1 to L2 message.

You can cancel a message after a predefined about of time from the moment this function is called. The amount of time is set by the link:https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/starknet/solidity/StarknetMessaging.sol#L59[`messageCancellationDelay()`] function.

> #Where can I see the actual numerical value?#

Reviewer: https://etherscan.io/address/0xc662c410c0ecf747543f5ba90660f6abebd9c8c4#readProxyContract#F11

[id="messaging_function_reference"]
= Messaging function and event reference
:description: Comprehensive function and event reference for the `StarknetMessaging` smart contracts.

> The `StarknetMessaging` smart contracts include functions and events that you use to implement various messaging flows in a dApp.

Reviewer 2: Good

> The `StarknetMessaging` smart contracts include functions and events that you use to implement various messaging flows in a dApp.

Author: @LucasLvy WDYT?

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L161[`cancelL1ToL2Message`]

'''

[#consumeMessageFromL2]
=== `consumeMessageFromL2`

[discrete]
==== Description

Consumes a message that was sent from an L2 contract.

Returns the hash of the message.

[discrete]
==== State Mutability

None.

[discrete]
==== Parameters

[horizontal,labelwidth="30",role=stripes-odd]
`uint256 _fromAddress_`:: The address of the L2 contract sending the message.
`uint256[] calldata _payload_`:: The payload of the message.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]
`bytes32 _msgHash_`:: The hash of the consumed message.

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L132[`consumeMessageFromL2`]

'''

[#getMaxL1MsgFee]
=== `getMaxL1MsgFee`

[discrete]
==== Description

> Returns the maximum fee, in Wei, that Starknet #charges# #(Is "charges" correct? It says "accepts" link:https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/starknet/solidity/IStarknetMessaging.sol#L8[in the source file])# for a single message.

Author: @LucasLvy Is "charges" correct? It says "accepts" [in the source file](https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/starknet/solidity/IStarknetMessaging.sol#L8).

Author: I spoke with Remo and changed accordingly.

> Returns the maximum fee, in Wei, that Starknet #charges# #(Is "charges" correct? It says "accepts" link:https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/starknet/solidity/IStarknetMessaging.sol#L8[in the source file])# for a single message.

Author: ```suggestion
Returns the maximum fee, in Wei, that Starknet accepts for a single message. If the fee passed is higher than this value, the transaction is not accepted.
```

[discrete]
==== State Mutability

`pure`

[discrete]
==== Parameters

None.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]
`uint256 _MAX_L1_MSG_FEE_`:: The maximum fee, in Wei, that Starknet #accepts/charges?# per single message.

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L31[`getMaxL1MsgFee`]

'''

[#l1ToL2MessageCancellations]
=== `l1ToL2MessageCancellations`

[discrete]
==== Description

> Returns the timestamp at the time cancelL1ToL2Message was called with a given message hash.

Author: ```suggestion
Returns the timestamp at the time `cancelL1ToL2Message` was called with a given message hash.
```

Reviewer 2: Good

Author: Based on Remo's input.

[discrete]
==== State Mutability

`view`

[discrete]
==== Parameters

[horizontal,labelwidth="30",role=stripes-odd]
`bytes32 _msgHash_`:: The message hash.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]

> `uint256 _result_`:: The timestamp at the time cancelL1ToL2Message was called with the given message hash `_msgHash_`. Returns 0 if cancelL1ToL2Message was never called with the given message hash `_msgHash_`.

Author: ```suggestion
`uint256 _result_`:: The timestamp, in seconds since the Linux epoch, at the time `cancelL1ToL2Message` was called with the message hash `_msgHash_`. 

Returns `0` if `cancelL1ToL2Message` was not called with the message hash `_msgHash_`.
```

Reviewer 2: Good

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L73[`l1ToL2MessageCancellations`]

'''

[#l1ToL2MessageNonce]
=== `l1ToL2MessageNonce`

[discrete]
==== Description

> Returns the nonce of the next message sending to L2 contract.

Author: @LucasLvy Does this mean that you first call this function and then send a message, let's call it snoopy, and the function returns the nonce of snoopy?

If the answer is _Yes_, then I would change this line to:

```suggestion
Returns the nonce of the next message sent to L2 contract.
```

Author: I'm clarifying with Remo.

[discrete]
==== State Mutability

`view`

[discrete]
==== Parameters

None.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]
`uint256 _nonce_`:: The nonce of the next message sending to L2 contract.

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L55[`l1ToL2MessageNonce`]

'''

[#l1ToL2Messages]
=== `l1ToL2Messages`

[discrete]
==== Description

Indicates if a pending message is associated with a given message hash.

[discrete]
==== State Mutability

`view`

[discrete]
==== Parameters

[horizontal,labelwidth="30",role=stripes-odd]
`bytes32 _msgHash_`:: The message hash.

[discrete]
==== Returns

Returns `uint256 _result_`, where `_result_` is one of the following:

> [horizontal,labelwidth="30",role=stripes-odd]

Author: ```suggestion
[horizontal,labelwidth="5",role=stripes-odd]
```

Reviewer 2: Good

> `msg_fee + 1`:: A pending message is associated with the `_msgHash_` parameter. #What is `msg_fee`? I don't see it mentioned anywhere. Does it simply refer to the fee of the pending message? Why does it return `msg_fee + 1`? How is that value used?#

Author: @LucasLvy Please respond.

Reviewer 2: // Note that the inclusion of the unique nonce in the message hash implies that
// l1ToL2Messages()[msgHash] was not accessed before.
l1ToL2Messages()[msgHash] = msg.value + 1;

refered from https://github.com/starkware-libs/cairo-lang/blob/efa9648f57568aad8f8a13fbf027d2de7c63c2c0/src/starkware/starknet/solidity/StarknetMessaging.sol#L121C1-L123C51

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L39[`l1ToL2Messages`]

'''

[#l2ToL1Messages]
=== `l2ToL1Messages`

[discrete]
==== Description

Indicates if a pending message is associated with a given message hash.

[discrete]
==== State Mutability

`view`

[discrete]
==== Parameters

[horizontal,labelwidth="30",role=stripes-odd]
`bytes32 _msgHash_`:: The message hash.

[discrete]
==== Returns

Returns `uint256 _result_`, where `_result_` is one of the following:
[horizontal,labelwidth="30",role=stripes-odd]
`1`:: A pending message is associated with the `_msgHash_` parameter.
`0`:: No pending message is associated with the `_msgHash_` parameter.

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L43[`l2ToL1Messages`]

'''

[#messageCancellationDelay]
=== `messageCancellationDelay`

[discrete]
==== Description

> Returns the time interval after which you can cancel a message starting from the moment of calling the xref:#startL1ToL2MessageCancellation[`startL1ToL2MessageCancellation`] function. You can get the real value by calling the link:https://etherscan.io/address/0xc662c410c0ecf747543f5ba90660f6abebd9c8c4#readProxyContract#F11[`messageCancellationDelay`] function on etherscan.

Author: ```suggestion
Returns the time interval after which you can cancel a message starting from the moment of calling the xref:#startL1ToL2MessageCancellation[`startL1ToL2MessageCancellation`] function. You can get the real value by calling the link:https://etherscan.io/address/0xc662c410c0ecf747543f5ba90660f6abebd9c8c4#readProxyContract#F11[`messageCancellationDelay`] function on a block explorer, such as Etherscan.
```

Reviewer 2: Good

'''

[#sendMessageToL2]
=== `sendMessageToL2`

[discrete]
==== Description

Sends a message to an L2 contract. The message fee is the cost of executing this function.

[discrete]
==== State Mutability

`payable`

[discrete]
==== Parameters

[horizontal,labelwidth="30",role=stripes-odd]
`uint256 _toAddress_`:: The address of the L2 contract.
`uint256 _selector_`::
include::messaging_reference.adoc[tag=uint256_selector]
`uint256[] calldata _payload_`:: The payload of the message.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]
`bytes32 _msgHash_`:: The hash of the message.
`uint256 _nonce_`:: The nonce of the message.

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L110[`sendMessageToL2`]

'''

[#startL1ToL2MessageCancellation]
=== `startL1ToL2MessageCancellation`

[discrete]
==== Description

> Starts the cancellation of an L1 to L2 message.

Author: @xiaolou86 Be consistent with how you refer to these messages. See next comment for more info.

Reviewer 2: ```suggestion
Starts the cancellation of a message from L1 to L2.
```

Reviewer 2: Done.

* Event: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol#L19[`ConsumedMessageToL1`]

'''

[#ConsumedMessageToL2]
=== `ConsumedMessageToL2`

[discrete]
==== Description

This event is emitted when a message from L1 to L2 is consumed.

[discrete]
==== Event attributes

[horizontal,role=stripes-odd]
`address indexed _fromAddress_`:: The address of the sender on L1.
`uint256 indexed _toAddress_`:: The address of the receiver on L2.
`uint256 indexed _selector_`::
include::messaging_reference.adoc[tag=uint256_selector]
`uint256[] _payload_`:: The payload of the consumed message.
`uint256 _nonce_`:: The nonce of the consumed message.

[discrete]
==== Event definition

* Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol[`IStarknetMessagingEvents`]

* Event: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol#L26[`ConsumedMessageToL2`]

'''

[#LogMessageToL1]
=== `LogMessageToL1`

[discrete]
==== Description

> This event is emitted when a message is sent from L2 to L1.

Reviewer 2: ```suggestion
This event is emitted when sending a message from L2 to L1.
```

[discrete]
==== Event attributes

[horizontal,role=stripes-odd]
`uint256 indexed _fromAddress_`:: The address of the sender on L2.
`address indexed _toAddress_`:: The address of the receiver on L1.
`uint256[] _payload_`:: The payload of the message.

[discrete]
==== Event definition

* Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol[`IStarknetMessagingEvents`]

* Event: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol#L6[`LogMessageToL1`]

'''

[#LogMessageToL2]
=== `LogMessageToL2`

[discrete]
==== Description

> This event is emitted when a message is sent from L1 to L2.

Author: @xiaolou86 Please be consistent with the way you refer to these messages. Let's use:

`...a message from L1 to L2` or `...a message from L2 to L1`.

Please check all occurrences and make them consistent.

Reviewer 2: ```suggestion
This event is emitted when sending a message from L1 to L2.
```

[discrete]
==== Event attributes

[horizontal,role=stripes-odd]
`address indexed _fromAddress_`:: The address of the sender on L1.
`uint256 indexed _toAddress_`:: The address of the receiver on L2.
`uint256 indexed _selector_`::
include::messaging_reference.adoc[tag=uint256_selector]
`uint256[] _payload_`:: The payload of the message.
`uint256 _nonce_`:: The nonce of the message.
`uint256 _fee_`:: The fee associated with the message.

[discrete]
==== Event definition

* Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol[`IStarknetMessagingEvents`]

* Event: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol#L9[`LogMessageToL2`]

'''

[#MessageToL2Canceled]
=== `MessageToL2Canceled`

[discrete]
==== Description

> This event is emitted when an L1 to L2 message is canceled.

Reviewer 2: ```suggestion
This event is emitted after calling the `cancelL1ToL2Message` function.
```

[discrete]
==== Event attributes

[horizontal,role=stripes-odd]
`address indexed _fromAddress_`:: The address of the sender on L1.
`uint256 indexed _toAddress_`:: The address of the receiver on L2.
`uint256 indexed _selector_`::
include::messaging_reference.adoc[tag=uint256_selector]
`uint256[] _payload_`:: The payload of the canceled message.
`uint256 _nonce_`:: The nonce of the canceled message.

[discrete]
==== Event definition

* Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol[`IStarknetMessagingEvents`]

* Event: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol#L44[`MessageToL2Canceled`]

'''

[#MessageToL2CancellationStarted]
=== `MessageToL2CancellationStarted`

[discrete]
==== Description

> This event is emitted when the cancellation of an L1 to L2 message is started.

Author: "is started" is awkward.

```suggestion
This event is emitted after calling the `startL1ToL2MessageCancellation` function.
```

Reviewer 2: Good

Consumes a message that was sent from an L2 contract.

Returns the hash of the message.

[discrete]
==== State Mutability

None.

[discrete]
==== Parameters

[horizontal,labelwidth="30",role=stripes-odd]
`uint256 _fromAddress_`:: The address of the L2 contract sending the message.
`uint256[] calldata _payload_`:: The payload of the message.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]
`bytes32 _msgHash_`:: The hash of the consumed message.

[discrete]
==== Function definition

Contract: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`]

* Function: link:https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol#L132[`consumeMessageFromL2`]

'''

[#getMaxL1MsgFee]
=== `getMaxL1MsgFee`

[discrete]
==== Description

Returns the maximum fee, in Wei, that Starknet accepts for a single message. If the fee passed is higher than this value, the transaction is not accepted.

[discrete]
==== State Mutability

`pure`

[discrete]
==== Parameters

None.

[discrete]
==== Returns

[horizontal,labelwidth="30",role=stripes-odd]

> `uint256 _MAX_L1_MSG_FEE_`:: The maximum fee, in Wei, that Starknet #accepts/charges?# per single message.

Author: ```suggestion
`uint256 _MAX_L1_MSG_FEE_`:: The maximum fee, in Wei, that Starknet accepts for a single message.
```
